scipy/stats/tests/test_morestats.py

```diff
@@ -2229,6 +2229,39 @@ def test_input_high_variance(self):
         xt_box, lam_box = stats.boxcox(x + 1)
         assert_allclose(xt_yeo, xt_box, rtol=1e-6)
         assert_allclose(lam_yeo, lam_box, rtol=1e-6)
+    
+    def test_overflow_expr(self):
+        # non-regression test for gh-18389
+        x = np.array([2003.0, 1950.0, 1997.0, 2000.0, 2009.0,
+                      2009.0, 1980.0, 1999.0, 2007.0, 1991.0])
+        
+        def optimizer(fun):
+            out = optimize.fminbound(fun, -lam_yeo, lam_yeo, xtol=1.48e-08)
+            result = optimize.OptimizeResult()
+            result.x = out
+            return result
+
+        with np.errstate(over="raise"):
+            xt_yeo, lam_yeo = stats.yeojohnson(x)
+            xt_box, lam_box = stats.boxcox(x + 1, optimizer=optimizer)
+            assert_allclose(lam_yeo, lam_box, rtol=1e-5)
+            assert_allclose(xt_yeo, xt_box, rtol=1e-5)
+    
+    def test_overflow_lam(self):
+        # non-regression test for gh-18389
+        x = np.array([2003.0, 1950.0, 1997.0, 2000.0, 2009.0])
```

Comment:
Can you just [@parametrize](https://docs.pytest.org/en/7.3.x/how-to/parametrize.html) the test above and comment on why these cases are different? (Do they test the two distinct sources of oveflow?)

Reply:
Yes, I wrote these as two different tests because they each test a different source of overflow. I just updated the tests as follows:
- The overflow tests are now parametrized as per your suggestion, and added a comment to explain what type of overflow they attempt to trigger.
- I added a test to ensure that the variance of the transformed data can be computed without overflow (cf. the commented implementation in my reply below).
- I made underflow also raise exceptions in case other test cases are added later that would trigger underflow.